### DISCUSSION TITLE
Rust WASM Runtime + WebSockets == 💔

### DISCUSSION THREAD

**Discussion by Markusdreyer** [Markusdreyer/pace-platform#3] [General]
Compiling Rust to a WASM target is not as straightforward as initially thought. The WASM binaries are very lightweight and barebones, and many of the commonly used libraries are [not built to support WASM targets](https://users.rust-lang.org/t/building-a-wasm-file-failed-due-to-unresolved-import/85291). This means that you currently need to build your WASM apps with a WASM-supported framework, significantly limiting the ecosystem of libraries to use. For example, building a WASM Websocket server is not supported out of the box in either [Fermyon](https://www.fermyon.com/) or [Cosmonic](https://cosmonic.com/).

**Comment by Markusdreyer** [comment]
The Cosmonic framework has [support for the NATS message broker](https://cosmonic.com/docs/capabilities/messaging/nats), [which supports WebSockets from version 2.2](https://docs.nats.io/running-a-nats-service/configuration/websocket).

**Comment by Markusdreyer** [comment]
A possible compromise to get _something_ out, but at the same time trying out WASM services could be to have the WebSocket server running as a traditional Docker image running in a conventional cloud service, and building the client API as a REST API, allowing for out-of-the-box WASM support with either Fermyon or Cosmonic.

**Reply by turbotobias** [reply]
Agree with REST, Docker and traditional cloud given wasm limitations. People like fly.io https://github.com/fly-apps/hello-rust, do you have a preference?

**Reply by Markusdreyer** [reply]
That's what I wanted to test out! Checking out this video to get a grasp of it: https://www.youtube.com/watch?v=gboGyccRVXI&t=14s

**Reply by Markusdreyer** [reply]
Whoops, not the right video... Can't find back to it now. Thought Jon Gjengset had a video on Fly.io
